pkg/cfn/builder/nodegroup.go

```diff
@@ -196,41 +184,115 @@ func (n *NodeGroupResourceSet) addResourcesForNodeGroup() error {
 			},
 		)
 	}
-	ngProps := map[string]interface{}{
-		"LaunchTemplate": map[string]interface{}{
-			"LaunchTemplateName": launchTemplateName,
-			"Version":            gfn.MakeFnGetAttString("NodeGroupLaunchTemplate.LatestVersionNumber"),
+
+	asg := nodeGroupResource(launchTemplateName, &vpcZoneIdentifier, tags, n.spec)
+	n.newResource("NodeGroup", asg)
+
+	return nil
+}
+
+// GetAllOutputs collects all outputs of the node group
+func (n *NodeGroupResourceSet) GetAllOutputs(stack cfn.Stack) error {
+	return n.rs.GetAllOutputs(stack)
+}
+
+func newLaunchTemplateData(n *NodeGroupResourceSet) *gfn.AWSEC2LaunchTemplate_LaunchTemplateData {
+	launchTemplateData := &gfn.AWSEC2LaunchTemplate_LaunchTemplateData{
+		IamInstanceProfile: &gfn.AWSEC2LaunchTemplate_IamInstanceProfile{
+			Arn: n.instanceProfileARN,
 		},
-		"VPCZoneIdentifier": vpcZoneIdentifier,
+		ImageId:      gfn.NewString(n.spec.AMI),
+		InstanceType: gfn.NewString(n.spec.InstanceType),
+		UserData:     n.userData,
+		NetworkInterfaces: []gfn.AWSEC2LaunchTemplate_NetworkInterface{{
+			AssociatePublicIpAddress: gfn.NewBoolean(!n.spec.PrivateNetworking),
+			DeviceIndex:              gfn.NewInteger(0),
+			Groups:                   n.securityGroups,
+		}},
+	}
+
+	return launchTemplateData
+}
+
+func nodeGroupResource(launchTemplateName *gfn.Value, vpcZoneIdentifier *interface{}, tags []map[string]interface{}, ng *api.NodeGroup) *awsCloudFormationResource {
+	ngProps := map[string]interface{}{
+		"VPCZoneIdentifier": *vpcZoneIdentifier,
 		"Tags":              tags,
 	}
-	if n.spec.DesiredCapacity != nil {
-		ngProps["DesiredCapacity"] = fmt.Sprintf("%d", *n.spec.DesiredCapacity)
+	if ng.DesiredCapacity != nil {
+		ngProps["DesiredCapacity"] = fmt.Sprintf("%d", *ng.DesiredCapacity)
 	}
-	if n.spec.MinSize != nil {
-		ngProps["MinSize"] = fmt.Sprintf("%d", *n.spec.MinSize)
+	if ng.MinSize != nil {
+		ngProps["MinSize"] = fmt.Sprintf("%d", *ng.MinSize)
 	}
-	if n.spec.MaxSize != nil {
-		ngProps["MaxSize"] = fmt.Sprintf("%d", *n.spec.MaxSize)
+	if ng.MaxSize != nil {
+		ngProps["MaxSize"] = fmt.Sprintf("%d", *ng.MaxSize)
 	}
-	if len(n.spec.TargetGroupARNs) > 0 {
-		ngProps["TargetGroupARNs"] = n.spec.TargetGroupARNs
+	if len(ng.TargetGroupARNs) > 0 {
+		ngProps["TargetGroupARNs"] = ng.TargetGroupARNs
+	}
+	if hasMixedInstances(ng) {
+		ngProps["MixedInstancesPolicy"] = *mixedInstancesPolicy(launchTemplateName, ng)
+	} else {
+		ngProps["LaunchTemplate"] = map[string]interface{}{
+			"LaunchTemplateName": launchTemplateName,
+			"Version":            gfn.MakeFnGetAttString("NodeGroupLaunchTemplate.LatestVersionNumber"),
+		}
 	}
-	n.newResource("NodeGroup", &awsCloudFormationResource{
+
+	return &awsCloudFormationResource{
 		Type:       "AWS::AutoScaling::AutoScalingGroup",
 		Properties: ngProps,
 		UpdatePolicy: map[string]map[string]string{
 			"AutoScalingRollingUpdate": {
-				"MinInstancesInService": "1",
+				"MinInstancesInService": "0",
```

Comment:
Any reason 0 is used here?

Reply:
This is what I've seen in some examples when people were using pure spot instance groups. I'm not sure what's best here. What do you think?

Reply:
0-downtime should be the goal. Seems you are right, 0 should be used because it's impossible for AWS to guarantee a minimum number of instances to be kept in service during the rolling update.

Reply:
I think I was one of the people who suggested this. Yes, this is *only* to be used for 100% spot instance node groups using launch _*configurations*_ (not templates) because CloudFormation does not support anything other than this config when updating LCs.